File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start = 0
        charSet = set()
        count = 0
        
        for end in range(len(s)):
            while s[end] in charSet:
                charSet.remove(s[start])
                start += 1
            
            charSet.add(s[end])
            count = max(count, end - start + 1)
        
        return count
        
    
    
    # have start and end pointer
    # create set to store chars iterated
    # iterate by index, if duplicate found in set, move start pointer, remove duplicate in set
    # else, add char in set and move end pointer

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let charSet = new Set();
    let start = 0;
    let count = 0;
    
    for (let end=0; end<s.length; end++) {
        while (charSet.has(s[end])) {
            charSet.delete(s[start]);
            start++;
        }
        charSet.add(s[end]);
        count = Math.max(count, end-start+1);
    }
    return count;
};


// initialize start pointer, count, charSet
// set while loop if end pointer is in set, then remove s[start] in set to move start pointer by start++
// otherwise, add char in set, check max (end-start+1)